### PR TITLE
 Implement toList and toListLazy for mutable arrays and vectors 

### PR DESCRIPTION
**Pull request by utdemir**
Depends on #179. As part of #165,

It turns out, there is two different flavours of `toList` functions.

* 'toList' returns a new list alongside with the array. In order to do that, it has to read the entire array before returning; this can not done lazily since the underlying mutable array might change.
* 'toListLazy' converts an array to a lazy list, consuming the array in the process. It can do it lazily, since it does not give the array back, therefore guaranteeing in won't be changed.

Vector's 'toList' and 'toListLazy' functions also have the same constraint above.

In order to implement the vector changes, it also adds `unsafeRead` and `unsafeWrite` functions to the vector interface.

### CODE REVIEW

File: src/Data/Array/Mutable/Linear.hs
Comment (aspiwack) [inline src/Data/Array/Mutable/Linear.hs:116]: Careful to rebase if the `Unrestricted -> Ur` change is merged before (and vice versa).

File: src/Data/Array/Mutable/Linear.hs
Comment (aspiwack) [inline src/Data/Array/Mutable/Linear.hs:194]: You probably want to explain this a little better.

Also, maybe this ought to be the default, and the other one the one with a long name. I strongly suspect that this is a more common case for `toList`: you do `toList` to have a lazy traversal of your array, and probably you are abandoning the array there. (the other one is “just” the composition of `dup2` and `toList` (which, by the way, is easier to express with the `_2` lens, maybe that's how it ought to be defined (when the `Dupable` instance is implemented)))

Comment (utdemir) [reply src/Data/Array/Mutable/Linear.hs:194]: > You probably want to explain this a little better.

Agreed, will do.

> ... the other one is “just” the composition of dup2 and toList ....

This is a great point, thank you; I completely agree. So, now my plan is to drop the strict function, make the lazy function the default. Once we have the dupable instance (awaiting #151), it'd be trivial to implement that.

I will do the same to the `Vector` interface too.

**Comment by utdemir** [comment]
I rebased this PR after `Ur` rename and made the default (shorter) `toList` method lazy as @aspiwack suggested. I did not add the old method as `toListStrict`, because I think it will be very trivial to implement in terms of `dup` and `toList`.

**Review by Divesh-Otwani — APPROVED** [review]
Looks good.

**Comment by utdemir** [comment]
After the merging of #187, I had to move the `toList` function to unlifted mutable arrays. Since obeying the linear API forced me to strictly consume the array:

```haskell
toList :: Array a #-> Ur [a]
toList (arr :: Array a) =
  size arr & \(arr', Ur s) ->
    go 0 s arr'
 where
  go :: Int -> Int -> Array a #-> Ur [a]
  go i len arr'
    | i == len = arr' `lseq` Ur []
    | otherwise =
        unsafeRead arr' i & \(arr'', Ur a) ->
          go (i+1) len arr'' & \(Ur xs) ->
            Ur (a : xs)
```

Here, _I think_, pattern matching on the recursive call before the `:` constructor would make us traverse the entire array before returning the first element.

However, I will merge in this change, because it is holding up a few others. Please, still do comment here if you think that approach is wrong, and I will address it.

**Comment by aspiwack** [comment]
> Here, _I think_, pattern matching on the recursive call before the `:` constructor would make us traverse the entire array before returning the first element.

It does sound accurate, indeed. I think that what we are seeing here is that in order to force an `Ur …`, we do have to consume every linear thingy it its closure. So it stands to reason that by the time we access the list, the array is consumed. Now, the only way we have of consuming an mutable array, currently, is to make it unavailable. So we kind of have to construct the entire list strictly.

Very handwavy, but I think that it's the idea.

But we can think this a bit further. There are really two `toList` that we may want to consider:
- A _linear_ `toList :: Array a #-> [Ur a]`. This one is lazy.
- The `toList` on immutable arrays: we can augment the arrays with an `O(1)` freezing function (like in the article), `Array a #-> Vector a` (or whatever immutable arrays are cf #185 ). Then `toList :: Array a -> Ur [a]; toList = fmap toList . freeze`.

Now, since the second one doesn't actually require a copy, it does subsume the first one. So probably we only need to do that one.

**Comment by utdemir** [comment]
Just to clarify, we currently have a lazy `toList :: Array a #-> Ur [a]` function; it's just implemented with `MutableArray#` primitives on unlifted arrays, and lifted arrays refer to that one. Howeve I think it should be safe since it consumes the array as you mentioned.

However, I do agree that a `O(1)` freeze function is what we need here. I _think_ we should be able to go from our arrays to a `Vector`, that'd definitely make a lot of things easier. I'll create an issue for it.

**Pull request merged**
